File: myFinance_improved_V2/lib/features/inventory_analysis/sales/domain/entities/sales_analytics.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'sales_analytics.freezed.dart';

/// Time range for analytics queries
enum TimeRange {
  today,
  past7Days,
  thisMonth,
  lastMonth,
  past90Days,
  thisYear,
  lastYear,
}

/// Grouping granularity
enum GroupBy {
  daily,
  weekly,
  monthly,
}

/// Analysis dimension
enum Dimension {
  total,
  category,
  brand,
  product,
}

/// Metric type
enum Metric {
  revenue,
  quantity,
  margin,
}

/// Extension for TimeRange
extension TimeRangeExtension on TimeRange {
  String get label {
    switch (this) {
      case TimeRange.today:
        return 'Today';
      case TimeRange.past7Days:
        return 'Past 7 Days';
      case TimeRange.thisMonth:
        return 'This Month';
      case TimeRange.lastMonth:
        return 'Last Month';
      case TimeRange.past90Days:
        return 'Past 90 Days';
      case TimeRange.thisYear:
        return 'This Year';
      case TimeRange.lastYear:
        return 'Last Year';
    }
  }

  (DateTime, DateTime) getDateRange() {
    final now = DateTime.now();
    final today = DateTime(now.year, now.month, now.day);

    switch (this) {
      case TimeRange.today:
        return (today, today);
      case TimeRange.past7Days:
        return (today.subtract(const Duration(days: 6)), today);
      case TimeRange.thisMonth:
        return (DateTime(now.year, now.month, 1), today);
      case TimeRange.lastMonth:
        final lastMonthEnd = DateTime(now.year, now.month, 0); // 지난달 말일
        final lastMonthStart = DateTime(lastMonthEnd.year, lastMonthEnd.month, 1);
        return (lastMonthStart, lastMonthEnd);
      case TimeRange.past90Days:
        return (today.subtract(const Duration(days: 89)), today);
      case TimeRange.thisYear:
        return (DateTime(now.year, 1, 1), today);
      case TimeRange.lastYear:
        return (DateTime(now.year - 1, 1, 1), DateTime(now.year - 1, 12, 31));
    }
  }

  GroupBy get recommendedGroupBy {
    switch (this) {
      case TimeRange.today:
      case TimeRange.past7Days:
        return GroupBy.daily;
      case TimeRange.thisMonth:
      case TimeRange.lastMonth:
        return GroupBy.daily;
      case TimeRange.past90Days:
        return GroupBy.weekly;
      case TimeRange.thisYear:
      case TimeRange.lastYear:
        return GroupBy.monthly;
    }
  }
}

/// Analytics data point - Pure Entity (NO JSON)
@freezed
class AnalyticsDataPoint with _$AnalyticsDataPoint {
  const AnalyticsDataPoint._();

  const factory AnalyticsDataPoint({
    required DateTime period,
    required String dimensionId,
    required String dimensionName,
    required double totalQuantity,
    required double totalRevenue,
    required double totalMargin,
    required double marginRate,
    required int invoiceCount,
    double? revenueGrowth,
    double? quantityGrowth,
    double? marginGrowth,
  }) = _AnalyticsDataPoint;
}

/// Analytics summary - Pure Entity (NO JSON)
@freezed
class AnalyticsSummary with _$AnalyticsSummary {
  const factory AnalyticsSummary({
    required double totalRevenue,
    required double totalQuantity,
    required double totalMargin,
    required double avgMarginRate,
    required int recordCount,
  }) = _AnalyticsSummary;
}

/// Complete analytics response - Pure Entity (NO JSON)
@freezed
class SalesAnalyticsResponse with _$SalesAnalyticsResponse {
  const SalesAnalyticsResponse._();

  const factory SalesAnalyticsResponse({
    required bool success,
    required AnalyticsSummary summary,
    required List<AnalyticsDataPoint> data,
    String? error,
  }) = _SalesAnalyticsResponse;

  bool get hasData => data.isNotEmpty;
}

/// Drill-down item - Pure Entity (NO JSON)
@freezed
class DrillDownItem with _$DrillDownItem {
  const DrillDownItem._();

  const factory DrillDownItem({
    required String id,
    required String name,
    required double totalQuantity,
    required double totalRevenue,
    required double totalMargin,
    required double marginRate,
    int? productCount,
    int? brandCount,
    String? categoryId,
    String? categoryName,
    String? brandId,
    String? brandName,
  }) = _DrillDownItem;
}

/// Drill-down response - Pure Entity (NO JSON)
@freezed
class DrillDownResponse with _$DrillDownResponse {
  const factory DrillDownResponse({
    required bool success,
    required String level,
    required List<DrillDownItem> data,
    String? parentId,
    String? error,
  }) = _DrillDownResponse;
}

/// Breadcrumb item for drill-down navigation
@freezed
class BreadcrumbItem with _$BreadcrumbItem {
  const factory BreadcrumbItem({
    String? id,
    required String name,
    required String level,
  }) = _BreadcrumbItem;
}

/// Drill-down state for navigation
@freezed
class DrillDownState with _$DrillDownState {
  const DrillDownState._();

  const factory DrillDownState({
    @Default([]) List<BreadcrumbItem> breadcrumbs,
    @Default('category') String currentLevel,
    String? parentId,
  }) = _DrillDownState;

  factory DrillDownState.initial() {
    return const DrillDownState(
      breadcrumbs: [BreadcrumbItem(id: null, name: 'All', level: 'category')],
      currentLevel: 'category',
      parentId: null,
    );
  }

  DrillDownState drillDown(String id, String name) {
    final nextLevel = currentLevel == 'category' ? 'brand' : 'product';
    return DrillDownState(
      breadcrumbs: [
        ...breadcrumbs,
        BreadcrumbItem(id: id, name: name, level: nextLevel)
      ],
      currentLevel: nextLevel,
      parentId: id,
    );
  }

  DrillDownState navigateTo(int index) {
    if (index >= breadcrumbs.length) return this;
    final newBreadcrumbs = breadcrumbs.sublist(0, index + 1);
    final targetItem = newBreadcrumbs.last;
    return DrillDownState(
      breadcrumbs: newBreadcrumbs,
      currentLevel: targetItem.level,
      parentId: targetItem.id,
    );
  }

  bool get canDrillDown => currentLevel != 'product';
  bool get canGoBack => breadcrumbs.length > 1;
}

File: website/src/features/inventory-analysis/domain/entities/salesAnalytics.ts
/**
 * SalesAnalytics Entity
 * Domain entity for sales analytics RPC response
 */

// Group by options
export type SalesGroupBy = 'daily' | 'weekly' | 'monthly' | 'total';

// Dimension options
export type SalesDimension = 'total' | 'category' | 'brand' | 'product';

// Metric options for sorting
export type SalesMetric = 'revenue' | 'quantity' | 'margin';

// Order direction
export type OrderDirection = 'ASC' | 'DESC';

// Analytics params
export interface SalesAnalyticsParams {
  startDate: string;
  endDate: string;
  groupBy: SalesGroupBy;
  dimension: SalesDimension;
  metric: SalesMetric;
  categoryId: string | null;
  timezone: string;
}

// Summary data
export interface SalesAnalyticsSummary {
  totalRevenue: number;
  totalQuantity: number;
  totalMargin: number;
  avgMarginRate: number;
  recordCount: number;
}

// Individual data item
export interface SalesAnalyticsItem {
  period: string;
  dimensionId: string;
  dimensionName: string;
  totalQuantity: number;
  totalRevenue: number;
  totalMargin: number;
  marginRate: number;
  invoiceCount: number;
  revenueGrowth: number | null;
  quantityGrowth: number | null;
  marginGrowth: number | null;
}

// Success response
export interface SalesAnalyticsSuccess {
  success: true;
  params: SalesAnalyticsParams;
  summary: SalesAnalyticsSummary;
  data: SalesAnalyticsItem[];
}

// Error response
export interface SalesAnalyticsError {
  success: false;
  error: string;
  detail: string;
}

// Union type
export type SalesAnalytics = SalesAnalyticsSuccess | SalesAnalyticsError;

// RPC response types
interface SalesAnalyticsParamsRpc {
  start_date: string;
  end_date: string;
  group_by: string;
  dimension: string;
  metric: string;
  category_id: string | null;
  timezone: string;
}

interface SalesAnalyticsSummaryRpc {
  total_revenue: number;
  total_quantity: number;
  total_margin: number;
  avg_margin_rate: number;
  record_count: number;
}

interface SalesAnalyticsItemRpc {
  period: string;
  dimension_id: string;
  dimension_name: string;
  total_quantity: number;
  total_revenue: number;
  total_margin: number;
  margin_rate: number;
  invoice_count: number;
  revenue_growth: number | null;
  quantity_growth: number | null;
  margin_growth: number | null;
}

interface SalesAnalyticsSuccessRpc {
  success: true;
  params: SalesAnalyticsParamsRpc;
  summary: SalesAnalyticsSummaryRpc;
  data: SalesAnalyticsItemRpc[];
}

interface SalesAnalyticsErrorRpc {
  success: false;
  error: string;
  detail: string;
}

type SalesAnalyticsRpc = SalesAnalyticsSuccessRpc | SalesAnalyticsErrorRpc;

/**
 * Map RPC response to domain entity
 */
export function mapSalesAnalyticsFromRpc(data: unknown): SalesAnalytics {
  const response = data as SalesAnalyticsRpc;

  if (!response.success) {
    return {
      success: false,
      error: response.error ?? 'Unknown error',
      detail: response.detail ?? '',
    };
  }

  return {
    success: true,
    params: {
      startDate: response.params?.start_date ?? '',
      endDate: response.params?.end_date ?? '',
      groupBy: (response.params?.group_by ?? 'monthly') as SalesGroupBy,
      dimension: (response.params?.dimension ?? 'total') as SalesDimension,
      metric: (response.params?.metric ?? 'revenue') as SalesMetric,
      categoryId: response.params?.category_id ?? null,
      timezone: response.params?.timezone ?? 'Asia/Ho_Chi_Minh',
    },
    summary: {
      totalRevenue: response.summary?.total_revenue ?? 0,
      totalQuantity: response.summary?.total_quantity ?? 0,
      totalMargin: response.summary?.total_margin ?? 0,
      avgMarginRate: response.summary?.avg_margin_rate ?? 0,
      recordCount: response.summary?.record_count ?? 0,
    },
    data: (response.data ?? []).map((item) => ({
      period: item.period ?? '',
      dimensionId: item.dimension_id ?? '',
      dimensionName: item.dimension_name ?? '',
      totalQuantity: item.total_quantity ?? 0,
      totalRevenue: item.total_revenue ?? 0,
      totalMargin: item.total_margin ?? 0,
      marginRate: item.margin_rate ?? 0,
      invoiceCount: item.invoice_count ?? 0,
      revenueGrowth: item.revenue_growth,
      quantityGrowth: item.quantity_growth,
      marginGrowth: item.margin_growth,
    })),
  };
}
